### PR TITLE
perf: Replace `jax.ops.segment_sum` with wider-accumulating `segment_sum` and use `segment_take` for float gather adjoints

### PR DESCRIPTION
**Pull request by n-gao**
This PR introduces a dedicated `segment_sum` utility and a new `segment_take` function, replacing direct calls to `jax.ops.segment_sum` throughout the codebase to improve numerical accuracy and gradient quality.

## Motivation

`jax.ops.segment_sum` accumulates contributions in the same float precision as the input, meaning a key receiving `k` contributions rounds `k` times. For float32 inputs with many contributions or large dynamic range, this causes meaningful numerical error. Similarly, the reverse pass of a plain `arr.at[idxs].get()` gather clamps out-of-bounds indices to the last row and piles cotangents there rather than distributing them correctly.

## Changes

**`segment_sum`**: The new utility accumulates in a wider float before casting back to the input dtype, so each key rounds once regardless of how many contributions it receives. Out-of-bounds sentinel indices contribute nothing.

**`segment_take`**: Replaces `arr.at[idxs].get()` for multi-row float leaves selected by a plain integer index array. Its reverse pass sums cotangents using the wider `segment_sum`, making the gather an exact adjoint of the sum. Out-of-range indices return a fill value (zero by default) rather than clamping to the last row.

The wide gather is applied selectively — only to multi-row, non-weakly-typed inexact leaves selected by a plain integer index array. Single-row tables, integer/boolean leaves, clamped-mode gathers, and non-row-selecting index expressions all retain plain `at[].get()` semantics.

## Tests

- `Index.sum_over` correctly resolves a cancelling spike that float32 sequential accumulation rounds away
- Out-of-bounds sentinel entries contribute nothing to any key
- Multi-dimensional indices group element-wise
- Reverse mode distributes each key's cotangent to its contributing rows
- `Table.__getitem__` gradient accumulates float32 cotangents in wider precision for rows gathered many times
- Single-row tables, non-float leaves, and non-row-selecting indices preserve existing plain indexing behavior
- Out-of-bounds reads return zero for float leaves while `Index` leaves retain their sentinel values

## Results

Lennard-Jones energy in float32 against a float64 reference, fcc Ar rattled by 0.05 Å:

| system | cutoff | before | after |
|---|---|---|---|
| 32 atoms | 10 Å | 7.03e-07 | 2.15e-09 (327x) |
| 108 atoms | 14 Å | 2.08e-05 | 4.64e-08 (449x) |

Per-row cotangent accumulation in float32 against a float64 reference, 64 rows of 3 features, maximum relative error:

| contributions per row | magnitude spread | plain gather | `segment_take` |
|---|---|---|---|
| 32 | none | 1.39e-07 | 4.12e-08 (3.4x) |
| 32 | none, 75% zeros | 9.02e-08 | 3.07e-08 (2.9x) |
| 256 | none | 7.46e-07 | 3.30e-08 (22.6x) |
| 256 | 4 decades | 3.81e-07 | 2.56e-08 (14.9x) |
| 2048 | 4 decades, 50% zeros | 1.03e-06 | 2.44e-08 (42.4x) |

Jitted Lennard-Jones energy and forces, arms compiled up front from one settled neighbour-list state and timed round-robin; median of 61 rounds on an L4, 41 on 8 pinned CPU cores:

| | 108 atoms | 500 atoms | 864 atoms |
|---|---|---|---|
| GPU before | 0.556 ms | 0.415 ms | 0.706 ms |
| GPU after | 0.556 ms (+0.1%) | 0.420 ms (+1.3%) | 0.704 ms (-0.2%) |
| CPU before | 15.47 ms | 15.56 ms | 61.44 ms |
| CPU after | 14.96 ms (-3.3%) | 13.50 ms (-13.3%) | 57.01 ms (-7.2%) |
| flops after | +0.8% | +5.3% | +8.7% |
| flops without the single-row guard | +34.2% | +44.2% | +57.2% |

Forces are bitwise unchanged; the energy gain is entirely `Index.sum_over`.

### CODE REVIEW

**Comment by n-gao** [comment]
* **#269** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/269?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#278** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/278?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#272** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/272?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/272?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#271** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/271?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#268** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/268?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#267** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/267?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#229** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/229?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#233** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/233?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#235** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/235?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#234** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/234?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#232** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/232?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#230** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/230?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#228** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/228?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#227** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/227?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#226** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/226?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#224** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/224?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#223** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/223?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#222** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/222?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#215** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/215?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#213** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/213?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#216](https://github.com/cusp-ai-oss/kUPS/pull/216) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/216?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#212** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/212?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#211** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/211?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#210** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/210?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#204** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/204?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#209** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/209?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#203** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/203?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#202** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/202?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#201** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/201?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#186** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/186?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#185** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/185?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#183** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/183?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#182** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/182?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#181** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/181?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#180** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/180?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#178** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/178?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#171** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/171?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#173](https://github.com/cusp-ai-oss/kUPS/pull/173) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/173?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#170** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/170?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#206** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/206?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#205** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/205?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by felixmusil — APPROVED** [review]
